gfx942 HHS/BBS/XFP32 Grouped GEMM FreeSize logic (NN,NT,TN,TT)

nakajee opened the pull request.

nakajee commented:
Converted from gfx942_80cu GridBased logic.
Steps to convert:
- GroupedGemm: false to true
- add "_GG_" (after Bias) in KernelNameMin and SolutionNameMin
- remove all size info
- add 2 lines with "- null" (after size info)
- change GridBased to FreeSize

nakajee commented:
ActivationFuncCall is converted to true (including changes in PR #983)

nakajee commented:
hipblaslt-bench verification (confirmed no issue with norm_check)

hipblaslt-bench -r f16_r --transA N --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f16_r --transA N --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f16_r --transA T --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f16_r --transA T --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r bf16_r --transA N --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r bf16_r --transA N --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r bf16_r --transA T --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r bf16_r --transA T --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f32_r --transA N --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --compute_type xf32_r --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f32_r --transA N --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --compute_type xf32_r --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f32_r --transA T --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --compute_type xf32_r --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f32_r --transA T --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --compute_type xf32_r --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info

jichangjichang commented:
For GG, it doesn't have size mapping concept in logic yaml. You should put those yamls into gfx942 but not just for 80cu only.

jichangjichang commented:
After enabling these for general gfx942, you should also need to add CI tests for these problem types for GG.

KKyang reviewed (changes requested):
Sorry, I didn't see there is no gtest added.

nakajee commented:
> For GG, it doesn't have size mapping concept in logic yaml. You should put those yamls into gfx942 but not just for 80cu only.

@jichangjichang , @KKyang 
I updated the files as you suggested
- added gtest for BBS/XF32 (HHS is already covered by existing gtest)
- merged HHS GG kernels into existing gfx942 GG FreeSize yamls
    - merge.py does not support FreeSize logic files. I manually merged them.
- added new BBS/XF32 GG kernels under gfx942/FreeSize (removed 80cu)

By the way, gtest failed with GSU kernel + GroupedGemm>1. I removed all GSU kernels to make it pass.

nakajee commented:
gtest passed on local gfx942 node
(1) all precheckin (including new gtest)
`
./build/release/clients/staging/hipblaslt-test --gtest_filter=*pre_checkin* | tee hipblaslt_test_precheckin.log
`
(2) new tests only
`
./build/release/clients/staging/hipblaslt-test --gtest_filter=*pre_checkin_matmul_groupedgemm_bf16* | tee hipblaslt_test_precheckin_groupedgemm_bf16.log
`
`
./build/release/clients/staging/hipblaslt-test --gtest_filter=*pre_checkin_matmul_groupedgemm_xf32* | tee hipblaslt_test_precheckin_groupedgemm_xf32.log
`

nakajee commented:
precheckin gtest failed because I committed old version of GG yaml (with GSU).
I reverted the previous one and re-committed the correct version (without GSU).

jichangjichang commented on clients/gtest/matmul_gtest.yaml:
```diff
@@ -881,6 +881,36 @@ Tests:
+  beta: [ 0.0, 2.0 ]
+  unit_check: 1
+  algo_method: [2]
+  gpu_arch: '94[0-2]'
```
you didn't put 942 yaml 
here should be only '942'

nakajee replied:
Thanks for your comment.
I changed to 942.

nakajee commented:
@jichangjichang @KKyang 
I ran hipblaslt-test (no option) on gfx942 80cu node and confirmed all tests passed.

[==========] 35372 tests from 13 test suites ran. (1821328 ms total)
[  PASSED  ] 35372 tests.

Please let me know if you have any further comment.

nakajee commented:
This is logic file update only and no change in tensilelite.
I did not run tensilelite test locally.

KKyang reviewed (approved):
@jichangjichang Good to go?

jichangjichang reviewed (approved):
lgtm, but I still have one question for step "merged HHS GG kernels into existing gfx942 GG FreeSize yamls".
gfx942 HHS already has a lot of GG solutions. Is there any special reason for that you need to merge more solutions for HHS GG?

nakajee commented:
> lgtm, but I still have one question for step "merged HHS GG kernels into existing gfx942 GG FreeSize yamls". gfx942 HHS already has a lot of GG solutions. Is there any special reason for that you need to merge more solutions for HHS GG?

That is mainly because best perf kernel tends to be different between gfx942 and gfx942 80cu node.
(for example, winners on 80cu tend to have MI32 and different GLVW compared to existing gfx942 kernels)
I saw 5-10% difference when I did tuning on gfx942 80cu node.
(this is about best perf. Average was much more different)

nakajee commented:
I would like to keep best perf kernels for both gfx942 and gfx942 80cu.

jichangjichang commented:
> > lgtm, but I still have one question for step "merged HHS GG kernels into existing gfx942 GG FreeSize yamls". gfx942 HHS already has a lot of GG solutions. Is there any special reason for that you need to merge more solutions for HHS GG?
> 
> That is mainly because best perf kernel tends to be different between gfx942 and gfx942 80cu node. (for example, winners on 80cu tend to have MI32 and different GLVW compared to existing gfx942 kernels) I saw 5-10% difference when I did tuning on gfx942 80cu node. (this is about best perf. Average was much more different)

Got it

nakajee commented:
All CI tests passed
